extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/UserData.java

```diff
@@ -0,0 +1,59 @@
+package io.quarkus.websockets.next;
+
+/**
+ * Mutable user data associated with a connection. Implementations must be thread-safe.
+ */
+public interface UserData {
+
+    /**
+     *
+     * @param <VALUE>
+     * @param key
+     * @return the value or {@code null} if no mapping is found
+     */
+    <VALUE> VALUE get(TypedKey<VALUE> key);
+
+    /**
+     * Associates the specified value with the specified key. An old value is replaced by the specified value.
+     *
+     * @param <ConnectionData.VALUE>
+     * @param key
+     * @param value
+     * @return the previous value associated with {@code key}, or {@code null} if no mapping exists
+     */
+    <VALUE> VALUE put(TypedKey<VALUE> key, VALUE value);
+
+    /**
+     *
+     * @param <VALUE>
+     * @param key
+     */
+    <VALUE> VALUE remove(TypedKey<VALUE> key);
+
+    int size();
+
+    void clear();
+
+    /**
+     * @param <TYPE> The type this key is used for.
+     */
+    record TypedKey<TYPE>(String value) {
+
+        public static TypedKey<Integer> forInt(String val) {
```

Comment:
Shouldn't it be an `int` here? 
```suggestion
        public static TypedKey<Integer> forInt(int val) {
```

Reply:
and then `return new TypedKey<>(String.valueOf(val));`?

Reply:
In other words, this factory method is used to create a typed key with the specified string value. I.e. you expect the value to be integer, but the key is still string-based.

Reply:
Can we hide this implementation detail by providing extra `put` and `get` methods in the `UserData`?

Reply:
It's actually not an implementation detail. We want the users to use the typed key in order to use generics to verify the type of the value in `put()` and to cast the value automatically when obtaining the value with `get()`. We could add `getString()` and `putString()` etc. but I don't think it would make the API more clear or useful. Also we would have to keep the generic way so that it could be used for "custom" types, similar to https://github.com/quarkusio/quarkus/pull/43787/files#diff-a63047e6c7bf732258f9858dcb9cb5303fa5ebe9756eb27cc970a4130190bdc3R62.